[incubator-kie-issues#1612] Fix BusinessCalendar behavior with inconsistent (full/partial) set of properties on `calendar.properties`

Abhitocode opened the pull request:
Modified BusinessCalendarImpl class to include validation for calendar properties preventing misconfigurations.

Made business.hours.per.day calculated value based on start hour and end hour.

Added validations for properties to throw exceptions when calendar.properties file is misconfigured.

Added tests in BusinessCalendarImplTest to verify the modified intialization logic.

Closes: https://github.com/apache/incubator-kie-issues/issues/1612

NOTE: Work in progress, modifying tests and adding new test in BusinessCalendarImplTest aligning the new changes w.r.t CaleandarBean.

kie-ci3 commented:
**PR job** `#1` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-runtimes -u https://github.com/apache/incubator-kie-kogito-runtimes/pull/3788 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3788/1/display/redirect

**Test results:**
- PASSED: 3056
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3788/1/testReport/org.kie.kogito.it/JDBCOptimisticLockingIT/PR_check___Build_projects___testAsyncWIH/">PR check / Build projects / org.kie.kogito.it.JDBCOptimisticLockingIT.testAsyncWIH</a></summary>
Assertion condition defined as a org.kie.kogito.it.PersistenceTest 1 expectation failed.<br/>Expected status code <404> but was <200>.<br/> within 10 seconds.
</details>

martinweiler commented on jbpm/jbpm-flow/src/main/java/org/jbpm/process/core/timer/BusinessCalendarImpl.java:
```diff
@@ -338,6 +336,12 @@ protected int getPropertyAsInt(String propertyName, String defaultValue) {
+    protected int getPropertyAsInt(String propertyName) {
+        String value = businessCalendarConfiguration.getProperty(propertyName);
+
+        return Integer.parseInt(value);
```
We might want to return a nicer error message if this fails:
```
Caused by: java.lang.NumberFormatException: For input string: ""
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
	at java.base/java.lang.Integer.parseInt(Integer.java:678)
	at java.base/java.lang.Integer.parseInt(Integer.java:786)
	at org.jbpm.process.core.timer.BusinessCalendarImpl.getPropertyAsInt(BusinessCalendarImpl.java:342)
```

martinweiler commented on jbpm/jbpm-flow/src/main/java/org/jbpm/process/core/timer/BusinessCalendarImpl.java:
```diff
@@ -494,4 +500,66 @@ protected void handleWeekend(Calendar c, boolean resetTime) {
+        if (!errorMessage.isEmpty()) {
+            throw new IllegalArgumentException(errorMessage.toString());
+        }
+    }
```
It would be good to emit a log message here that summarizes the settings, including the configured ones.

martinweiler commented on jbpm/jbpm-flow/src/main/java/org/jbpm/process/core/timer/BusinessCalendarImpl.java:
```diff
@@ -90,15 +90,14 @@ public class BusinessCalendarImpl implements BusinessCalendar {
     private static final int SIM_SEC = 11;
 
-    public static final String DAYS_PER_WEEK = "business.days.per.week";
-    public static final String HOURS_PER_DAY = "business.hours.per.day";
```
Make sure the javadoc of this class is in-line with these changes (removed properties, removed default values)

jomarko reviewed (commented):
@Abhitocode hi, as I reviewed https://github.com/apache/incubator-kie-kogito-runtimes/pull/3795 and this PR is related, I share feedback also on this PR.

### 01
As new potential contributor to runtimes and calendars, reading the code:
```
businessCal.rollCalendarAfterHolidays(calendar, false);
```

It is difficult to understand the `false`. I mean, without checking the actual method code, it is not clear what that false means. I understand that author of the code may not have this problem, but new contributors may be in similar situation as me.

What do you think about introducing two separate API methods:
- `businessCal.rollCalendarAfterHolidays(calendar); // businessCal.rollCalendarAfterHolidays(calendar, false);`
- `businessCal.rollCalendarAfterHolidaysAndResetTime(calendar); //businessCal.rollCalendarAfterHolidays(calendar, true);`

or maybe change the boolean to some enum?
- `businessCal.rollCalendarAfterHolidays(calendar, RESET_TIMER.NO);`

### 02
```
    @Test
    void rollCalendarToNextWorkingDay() {
        List<Integer> workingDays = IntStream.range(Calendar.MONDAY, Calendar.SATURDAY).boxed().toList();
        List<Integer> weekendDays = Arrays.asList(Calendar.SATURDAY, Calendar.SUNDAY);
        Properties config = new Properties();
        config.setProperty(START_HOUR, "9");
        config.setProperty(END_HOUR, "17");
        config.setProperty(WEEKEND_DAYS, weekendDays.stream().map(String::valueOf).collect(Collectors.joining(",")));
        BusinessCalendarImpl businessCal = BusinessCalendarImpl.builder().withCalendarBean(new CalendarBean(config)).build();

        workingDays.forEach(workingDay -> {
            Calendar calendar = getCalendarAtExpectedWeekDay(workingDay);
            businessCal.rollCalendarToNextWorkingDay(calendar, false);
            assertThat(calendar.get(Calendar.DAY_OF_WEEK)).isEqualTo(workingDay);
        });
        weekendDays.forEach(weekendDay -> {
            Calendar calendar = getCalendarAtExpectedWeekDay(weekendDay);
            businessCal.rollCalendarToNextWorkingDay(calendar, false);
            assertThat(calendar.get(Calendar.DAY_OF_WEEK)).isEqualTo(Calendar.MONDAY);
        });
    }
```    

from reading this test ^, is `rollCalendarToNextWorkingDay` actually rolling? It seems we roll `workingDay` and after rolling we compare for equality with the original value. I would expect if we roll, that new value is `+1`. I mean I would expect sth like:
```
workingDays.forEach(workingDay -> {
            Calendar calendar = getCalendarAtExpectedWeekDay(workingDay);
            businessCal.rollCalendarToNextWorkingDay(calendar, false);
            assertThat(calendar.get(Calendar.DAY_OF_WEEK)).isEqualTo(workingDay + 1);
        });
```

Sorry if my feedback is not useful, or it proves my lack of knowledge about the domain. Thank you for your work.

Abhitocode commented:
Hi @jomarko, thank you so much for the feedback, for I have introduced variable for the same, I think it should be bit meaningful.
Also, I have updated the test name to rollCalendarToNextWorkingDayIfCurrentDayisNonWorking, it will roll to next working day if current day is non working. Please let me know if any other changes need to be addressed : )

pibizza commented on jbpm/jbpm-flow/src/test/java/org/jbpm/process/core/timer/CalendarBeanTest.java:
```diff
@@ -0,0 +1,295 @@
+        calendarConfiguration.put(propertyName, value);
+        int retrieved = CalendarBean.getPropertyAsInt(propertyName, calendarConfiguration);
+        assertThat(retrieved).isEqualTo(originalValue);
+        value = "WRONG";
```
Do not reuse variables in tests. they tend to make the test more confused.

pibizza commented on jbpm/jbpm-flow/src/test/java/org/jbpm/process/core/timer/CalendarBeanTest.java:
```diff
@@ -0,0 +1,295 @@
+        retrieved = CalendarBean.getSimpleDateFormat("dd-MM-yyyy");
+        assertThat(retrieved).isNotNull();
+        String wrong = "WRONG";
+        IllegalArgumentException thrown = assertThrows(IllegalArgumentException.class, () -> CalendarBean.getSimpleDateFormat(wrong));
```
do not use assertions from JUnit5. Use assertions from assertj. This is the standard in Kie.

pibizza commented on jbpm/jbpm-flow/src/test/java/org/jbpm/process/core/timer/CalendarBeanTest.java:
```diff
@@ -0,0 +1,295 @@
+        assertThat(retrievedErrors).contains(retrievedErrors);
+    }
+
+    private void commonIllegalArgumentAssertion(Executable executedMethod, List<String> errorMessages) {
```
Hiding assertions inside general methods tend to make the test less clear. Ideally assertions should be in the main body of the test itself.

gitgabrio replied:
HI @pibizza 
While I agree in general term, in that specific case there is a pretty extended setup for each method, and that setup could also be changed/adapted, in future.
Having to rewrite it for each method would make the overall class harder to maintain

pibizza replied:
This method is two lines, and probably can be collapsed in one assertion. There is no evident need for this specific method.

pibizza replied:
Sure, no worries.

pibizza reviewed (changes requested):
Some cleaning up for tests.

kie-ci3 commented:
**PR job** `#12` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-runtimes -u https://github.com/apache/incubator-kie-kogito-runtimes/pull/3788 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3788/12/display/redirect

**Test results:**
- PASSED: 3483
- FAILED: 2

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3788/12/testReport/org.kie.kogito.codegen.tests/PublishEventIT/testBasicUserTaskProcessAbort/">org.kie.kogito.codegen.tests.PublishEventIT.testBasicUserTaskProcessAbort</a></summary>
<br/>Expected size: 3 but was: 14 in:<br/>[ProcessInstanceStateDataEvent {specVersion=1.0, id='f5d01e62-6dd3-4b8c-8cbb-98fe6f2420db', source=http://myhost/TimerProcess, type='ProcessInstanceStateDataEvent', time=2024-12-02T03:14:21.601700977-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=ProcessInstanceStateEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, eventType=1, processId=defaultPackage.TimerProcess, processVersion=1.0, processType=BPMN, processInstanceId=1a5d9d4f-bd52-48cc-a39c-a0679227585c, businessKey=null, processName=TimerProcess, parentInstanceId=null, rootProcessId=null, rootProcessInstanceId=null, state=1, roles=null], kogitoProcessInstanceId='1a5d9d4f-bd52-48cc-a39c-a0679227585c', kogitoRootProcessInstanceId='null', kogitoProcessId='defaultPackage.TimerProcess', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitoproctype=BPMN, kogitoprocinstanceid=1a5d9d4f-bd52-48cc-a39c-a0679227585c, kogitoprocist=1, kogitoprocversion=1.0, kogitoprocid=defaultPackage.TimerProcess, kogitoaddons=test}},<br/>    ProcessInstanceNodeDataEvent {specVersion=1.0, id='304320b6-7193-46c8-a78e-e4201d4fa9f5', source=http://myhost/TimerProcess, type='ProcessInstanceNodeDataEvent', time=2024-12-02T03:14:21.595323682-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, eventType=1, processId=defaultPackage.TimerProcess, processVersion=1.0, processInstanceId=1a5d9d4f-bd52-48cc-a39c-a0679227585c, connectionNodeInstanceId=null, nodeDefinitionId=_F9AC8B12-8E75-49B5-B799-292F66270627, nodeName=start, nodeType=StartNode, nodeInstanceId=9c5e7109-b611-406e-8842-df84a919778f, data={}], kogitoProcessInstanceId='1a5d9d4f-bd52-48cc-a39c-a0679227585c', kogitoRootProcessInstanceId='null', kogitoProcessId='defaultPackage.TimerProcess', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitoproctype=BPMN, kogitoprocinstanceid=1a5d9d4f-bd52-48cc-a39c-a0679227585c, kogitoprocist=1, kogitoprocversion=1.0, kogitoprocid=defaultPackage.TimerProcess, kogitoaddons=test}},<br/>    ProcessInstanceNodeDataEvent {specVersion=1.0, id='2606a4e5-2f9e-471d-85ac-12706a974f71', source=http://myhost/TimerProcess, type='ProcessInstanceNodeDataEvent', time=2024-12-02T03:14:21.595734569-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, eventType=1, processId=defaultPackage.TimerProcess, processVersion=1.0, processInstanceId=1a5d9d4f-bd52-48cc-a39c-a0679227585c, connectionNodeInstanceId=SequenceFlow_2, nodeDefinitionId=_E76AD186-6FA6-4378-B45B-C8F73E2C497C, nodeName=timer fired, nodeType=ActionNode, nodeInstanceId=9540f5d3-4d4c-4059-883c-7f5195980341, data={}], kogitoProcessInstanceId='1a5d9d4f-bd52-48cc-a39c-a0679227585c', kogitoRootProcessInstanceId='null', kogitoProcessId='defaultPackage.TimerProcess', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitoproctype=BPMN, kogitoprocinstanceid=1a5d9d4f-bd52-48cc-a39c-a0679227585c, kogitoprocist=1, kogitoprocversion=1.0, kogitoprocid=defaultPackage.TimerProcess, kogitoaddons=test}},<br/>    ProcessInstanceNodeDataEvent {specVersion=1.0, id='1efef96f-b6c8-4d04-9f5e-357ff9363003', source=http://myhost/TimerProcess, type='ProcessInstanceNodeDataEvent', time=2024-12-02T03:14:21.596212269-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, eventType=1, processId=defaultPackage.TimerProcess, processVersion=1.0, processInstanceId=1a5d9d4f-bd52-48cc-a39c-a0679227585c, connectionNodeInstanceId=SequenceFlow_1, nodeDefinitionId=UserTask_1, nodeName=User Task 1, nodeType=HumanTaskNode, nodeInstanceId=b9104ded-e04b-4ae5-8a78-349856fa661d, data={}], kogitoProcessInstanceId='1a5d9d4f-bd52-48cc-a39c-a0679227585c', kogitoRootProcessInstanceId='null', kogitoProcessId='defaultPackage.TimerProcess', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitoproctype=BPMN, kogitoprocinstanceid=1a5d9d4f-bd52-48cc-a39c-a0679227585c, kogitoprocist=1, kogitoprocversion=1.0, kogitoprocid=defaultPackage.TimerProcess, kogitoaddons=test}},<br/>    UserTaskInstanceStateDataEvent {specVersion=1.0, id='3ce2e0da-1bdf-4945-85fa-38be6a4c174a', source=/process/null, type='UserTaskInstanceStateDataEvent', time=2024-12-02T03:14:21.597620472-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=UserTaskInstanceStateEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, userTaskDefinitionId=UserTask_1, userTaskInstanceId=dd152003-0656-49ea-a354-a20fa83ab78d, userTaskName=Task Name, userTaskDescription=null, userTaskPriority=1, userTaskReferenceName=User Task 1, state=Created, actualOwner=null, processInstanceId=null], kogitoProcessInstanceId='null', kogitoRootProcessInstanceId='null', kogitoProcessId='null', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitousertaskist=Created, kogitoprocrefid=User Task 1, kogitoprocist=null, kogitoaddons=test, kogitousertaskiid=dd152003-0656-49ea-a354-a20fa83ab78d}},<br/>    UserTaskInstanceStateDataEvent {specVersion=1.0, id='7baaff1c-8cea-4ce9-95c1-0aaee26a7cde', source=/process/null, type='UserTaskInstanceStateDataEvent', time=2024-12-02T03:14:21.597956452-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=UserTaskInstanceStateEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, userTaskDefinitionId=UserTask_1, userTaskInstanceId=dd152003-0656-49ea-a354-a20fa83ab78d, userTaskName=Task Name, userTaskDescription=null, userTaskPriority=1, userTaskReferenceName=User Task 1, state=Created, actualOwner=null, processInstanceId=null], kogitoProcessInstanceId='null', kogitoRootProcessInstanceId='null', kogitoProcessId='null', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitousertaskist=Created, kogitoprocrefid=User Task 1, kogitoprocist=null, kogitoaddons=test, kogitousertaskiid=dd152003-0656-49ea-a354-a20fa83ab78d}},<br/>    UserTaskInstanceStateDataEvent {specVersion=1.0, id='3ea216c1-7503-42c2-9433-e013d8d67c77', source=/process/null, type='UserTaskInstanceStateDataEvent', time=2024-12-02T03:14:21.598285652-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=UserTaskInstanceStateEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, userTaskDefinitionId=UserTask_1, userTaskInstanceId=dd152003-0656-49ea-a354-a20fa83ab78d, userTaskName=Task Name, userTaskDescription=null, userTaskPriority=1, userTaskReferenceName=User Task 1, state=Created, actualOwner=null, processInstanceId=null], kogitoProcessInstanceId='null', kogitoRootProcessInstanceId='null', kogitoProcessId='null', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitousertaskist=Created, kogitoprocrefid=User Task 1, kogitoprocist=null, kogitoaddons=test, kogitousertaskiid=dd152003-0656-49ea-a354-a20fa83ab78d}},<br/>    UserTaskInstanceStateDataEvent {specVersion=1.0, id='96a35c04-de45-4c11-bf22-88b62c7949c1', source=http://myhost/TimerProcess, type='UserTaskInstanceStateDataEvent', time=2024-12-02T03:14:21.599720327-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=UserTaskInstanceStateEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, userTaskDefinitionId=UserTask_1, userTaskInstanceId=dd152003-0656-49ea-a354-a20fa83ab78d, userTaskName=Task Name, userTaskDescription=null, userTaskPriority=1, userTaskReferenceName=User Task 1, state=Created, actualOwner=null, processInstanceId=1a5d9d4f-bd52-48cc-a39c-a0679227585c], kogitoProcessInstanceId='1a5d9d4f-bd52-48cc-a39c-a0679227585c', kogitoRootProcessInstanceId='null', kogitoProcessId='defaultPackage.TimerProcess', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitousertaskist=Created, kogitoprocrefid=User Task 1, kogitoproctype=BPMN, kogitoprocinstanceid=1a5d9d4f-bd52-48cc-a39c-a0679227585c, kogitoprocist=1, kogitoprocversion=1.0, kogitoprocid=defaultPackage.TimerProcess, kogitoaddons=test, kogitousertaskiid=dd152003-0656-49ea-a354-a20fa83ab78d}},<br/>    UserTaskInstanceStateDataEvent {specVersion=1.0, id='c668e2cb-f264-4308-9e6c-0bfeaaa7dee3', source=http://myhost/TimerProcess, type='UserTaskInstanceStateDataEvent', time=2024-12-02T03:14:21.600725086-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=UserTaskInstanceStateEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, userTaskDefinitionId=UserTask_1, userTaskInstanceId=dd152003-0656-49ea-a354-a20fa83ab78d, userTaskName=Task Name, userTaskDescription=null, userTaskPriority=1, userTaskReferenceName=User Task 1, state=Ready, actualOwner=null, processInstanceId=1a5d9d4f-bd52-48cc-a39c-a0679227585c], kogitoProcessInstanceId='1a5d9d4f-bd52-48cc-a39c-a0679227585c', kogitoRootProcessInstanceId='null', kogitoProcessId='defaultPackage.TimerProcess', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitousertaskist=Ready, kogitoprocrefid=User Task 1, kogitoproctype=BPMN, kogitoprocinstanceid=1a5d9d4f-bd52-48cc-a39c-a0679227585c, kogitoprocist=1, kogitoprocversion=1.0, kogitoprocid=defaultPackage.TimerProcess, kogitoaddons=test, kogitousertaskiid=dd152003-0656-49ea-a354-a20fa83ab78d}},<br/>    ProcessInstanceNodeDataEvent {specVersion=1.0, id='0cc1ba63-022f-4d98-b98a-77af27c81f3a', source=http://myhost/TimerProcess, type='ProcessInstanceNodeDataEvent', time=2024-12-02T03:14:21.601273808-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, eventType=2, processId=defaultPackage.TimerProcess, processVersion=1.0, processInstanceId=1a5d9d4f-bd52-48cc-a39c-a0679227585c, connectionNodeInstanceId=SequenceFlow_1, nodeDefinitionId=_E76AD186-6FA6-4378-B45B-C8F73E2C497C, nodeName=timer fired, nodeType=ActionNode, nodeInstanceId=9540f5d3-4d4c-4059-883c-7f5195980341, data={}], kogitoProcessInstanceId='1a5d9d4f-bd52-48cc-a39c-a0679227585c', kogitoRootProcessInstanceId='null', kogitoProcessId='defaultPackage.TimerProcess', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitoproctype=BPMN, kogitoprocinstanceid=1a5d9d4f-bd52-48cc-a39c-a0679227585c, kogitoprocist=1, kogitoprocversion=1.0, kogitoprocid=defaultPackage.TimerProcess, kogitoaddons=test}},<br/>    ProcessInstanceNodeDataEvent {specVersion=1.0, id='53c8fde4-c985-40a7-ba40-25865c46b7b4', source=http://myhost/TimerProcess, type='ProcessInstanceNodeDataEvent', time=2024-12-02T03:14:21.601486216-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, eventType=2, processId=defaultPackage.TimerProcess, processVersion=1.0, processInstanceId=1a5d9d4f-bd52-48cc-a39c-a0679227585c, connectionNodeInstanceId=SequenceFlow_2, nodeDefinitionId=_F9AC8B12-8E75-49B5-B799-292F66270627, nodeName=start, nodeType=StartNode, nodeInstanceId=9c5e7109-b611-406e-8842-df84a919778f, data={}], kogitoProcessInstanceId='1a5d9d4f-bd52-48cc-a39c-a0679227585c', kogitoRootProcessInstanceId='null', kogitoProcessId='defaultPackage.TimerProcess', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitoproctype=BPMN, kogitoprocinstanceid=1a5d9d4f-bd52-48cc-a39c-a0679227585c, kogitoprocist=1, kogitoprocversion=1.0, kogitoprocid=defaultPackage.TimerProcess, kogitoaddons=test}},<br/>    UserTaskInstanceStateDataEvent {specVersion=1.0, id='a0c12caa-d54b-46f9-a0d7-b771d39d63ea', source=http://myhost/UserTasksProcess, type='UserTaskInstanceStateDataEvent', time=2024-12-02T03:14:21.613150161-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=UserTaskInstanceStateEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, userTaskDefinitionId=UserTask_1, userTaskInstanceId=4ccf8635-2d3b-4e55-b0b9-162da8a65fb7, userTaskName=FirstTask, userTaskDescription=null, userTaskPriority=1, userTaskReferenceName=First Task, state=Error, actualOwner=null, processInstanceId=f44dbb88-8d02-47ce-88a9-22cd81fa987b], kogitoProcessInstanceId='f44dbb88-8d02-47ce-88a9-22cd81fa987b', kogitoRootProcessInstanceId='null', kogitoProcessId='UserTasksProcess', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitousertaskist=Error, kogitoprocrefid=First Task, kogitoproctype=BPMN, kogitoprocinstanceid=f44dbb88-8d02-47ce-88a9-22cd81fa987b, kogitoprocist=1, kogitoprocversion=1.0, kogitoprocid=UserTasksProcess, kogitoaddons=test, kogitousertaskiid=4ccf8635-2d3b-4e55-b0b9-162da8a65fb7}},<br/>    ProcessInstanceNodeDataEvent {specVersion=1.0, id='ace52821-373d-425f-90e7-542c0ef66e0c', source=http://myhost/UserTasksProcess, type='ProcessInstanceNodeDataEvent', time=2024-12-02T03:14:21.613786065-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, eventType=3, processId=UserTasksProcess, processVersion=1.0, processInstanceId=f44dbb88-8d02-47ce-88a9-22cd81fa987b, connectionNodeInstanceId=null, nodeDefinitionId=UserTask_1, nodeName=First Task, nodeType=HumanTaskNode, nodeInstanceId=84d75144-0182-4bdf-a2a1-f045609a4339, data={WorkItemId=e6e6b1f3-8849-4db8-af32-373f7e397510, WorkItemExternalReferenceId=4ccf8635-2d3b-4e55-b0b9-162da8a65fb7}], kogitoProcessInstanceId='f44dbb88-8d02-47ce-88a9-22cd81fa987b', kogitoRootProcessInstanceId='null', kogitoProcessId='UserTasksProcess', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitoproctype=BPMN, kogitoprocinstanceid=f44dbb88-8d02-47ce-88a9-22cd81fa987b, kogitoprocist=3, kogitoprocversion=1.0, kogitoprocid=UserTasksProcess, kogitoaddons=test}},<br/>    ProcessInstanceStateDataEvent {specVersion=1.0, id='e9826e9d-a8d4-4688-8cf2-d91069d875a9', source=http://myhost/UserTasksProcess, type='ProcessInstanceStateDataEvent', time=2024-12-02T03:14:21.614452911-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=ProcessInstanceStateEventBody [eventDate=Mon Dec 02 03:14:21 EST 2024, eventUser=null, eventType=2, processId=UserTasksProcess, processVersion=1.0, processType=BPMN, processInstanceId=f44dbb88-8d02-47ce-88a9-22cd81fa987b, businessKey=null, processName=UserTasksProcess, parentInstanceId=null, rootProcessId=null, rootProcessInstanceId=null, state=3, roles=null], kogitoProcessInstanceId='f44dbb88-8d02-47ce-88a9-22cd81fa987b', kogitoRootProcessInstanceId='null', kogitoProcessId='UserTasksProcess', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitoproctype=BPMN, kogitoprocinstanceid=f44dbb88-8d02-47ce-88a9-22cd81fa987b, kogitoprocist=3, kogitoprocversion=1.0, kogitoprocid=UserTasksProcess, kogitoaddons=test}}]
</details>
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3788/12/testReport/org.kie.kogito.codegen.tests/PublishEventIT/testCompensationProcess/">org.kie.kogito.codegen.tests.PublishEventIT.testCompensationProcess</a></summary>
<br/>Expected size: 9 but was: 11 in:<br/>[ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:23 EST 2024, eventUser=null, eventType=2, processId=defaultPackage.TimerProcess, processVersion=1.0, processInstanceId=6f8caf6f-3389-4f11-9d8a-aafe5285f821, connectionNodeInstanceId=SequenceFlow_1, nodeDefinitionId=_E76AD186-6FA6-4378-B45B-C8F73E2C497C, nodeName=timer fired, nodeType=ActionNode, nodeInstanceId=af72f81a-921a-4e4b-ab9b-77df6bd51539, data={}],<br/>    ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:23 EST 2024, eventUser=null, eventType=2, processId=defaultPackage.TimerProcess, processVersion=1.0, processInstanceId=6f8caf6f-3389-4f11-9d8a-aafe5285f821, connectionNodeInstanceId=SequenceFlow_2, nodeDefinitionId=_F9AC8B12-8E75-49B5-B799-292F66270627, nodeName=start, nodeType=StartNode, nodeInstanceId=cc551f02-feb4-43cb-8e63-75c4792a065b, data={}],<br/>    ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:23 EST 2024, eventUser=null, eventType=2, processId=compensateAll, processVersion=1.0, processInstanceId=5e4847e5-1990-4d78-9368-c292d7f66c5f, connectionNodeInstanceId=null, nodeDefinitionId=_B580FC2E-3924-472A-BCFB-A2ACBADA2B94, nodeName=Compensation 2, nodeType=ActionNode, nodeInstanceId=03755a57-0d5e-4cc1-ae60-0a97acb2cb1d, data={}],<br/>    ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:23 EST 2024, eventUser=null, eventType=2, processId=compensateAll, processVersion=1.0, processInstanceId=5e4847e5-1990-4d78-9368-c292d7f66c5f, connectionNodeInstanceId=, nodeDefinitionId=_3F8F7865-7B32-4E44-B323-44CCBFC96C8F, nodeName=BoundaryEvent, nodeType=BoundaryEventNode, nodeInstanceId=cf9e3f89-d788-47ba-ab56-faa13b897706, data={}],<br/>    ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:23 EST 2024, eventUser=null, eventType=2, processId=compensateAll, processVersion=1.0, processInstanceId=5e4847e5-1990-4d78-9368-c292d7f66c5f, connectionNodeInstanceId=null, nodeDefinitionId=_D91EB7E7-A287-4958-BF8E-27E364349BE9, nodeName=Compensation, nodeType=ActionNode, nodeInstanceId=9af6f2f4-6f5a-4d77-a908-59d307bd2112, data={}],<br/>    ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:23 EST 2024, eventUser=null, eventType=2, processId=compensateAll, processVersion=1.0, processInstanceId=5e4847e5-1990-4d78-9368-c292d7f66c5f, connectionNodeInstanceId=, nodeDefinitionId=_C9728638-4D98-4C28-972D-C8C5A7CA0290, nodeName=BoundaryEvent, nodeType=BoundaryEventNode, nodeInstanceId=83f72398-326e-474a-8107-2b0335497e20, data={}],<br/>    ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:23 EST 2024, eventUser=null, eventType=2, processId=compensateAll, processVersion=1.0, processInstanceId=5e4847e5-1990-4d78-9368-c292d7f66c5f, connectionNodeInstanceId=null, nodeDefinitionId=_E38F40D5-3A9A-4332-A4DC-CE0A2D37E179, nodeName=End, nodeType=EndNode, nodeInstanceId=d9e67cbe-c32a-44dc-b19c-a394719d20a4, data={}],<br/>    ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:23 EST 2024, eventUser=null, eventType=2, processId=compensateAll, processVersion=1.0, processInstanceId=5e4847e5-1990-4d78-9368-c292d7f66c5f, connectionNodeInstanceId=_457153B3-F67D-4D9F-ABF8-FFD10B6D934C, nodeDefinitionId=_CC9F5867-27CF-42CF-8838-648F44FD4CF1, nodeName=Compensate all tasks, nodeType=ActionNode, nodeInstanceId=efef756b-d8ce-4313-bec7-9148859d8d53, data={}],<br/>    ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:23 EST 2024, eventUser=null, eventType=2, processId=compensateAll, processVersion=1.0, processInstanceId=5e4847e5-1990-4d78-9368-c292d7f66c5f, connectionNodeInstanceId=_ACA65ACF-CBDA-4452-8C1A-609D54F1B749, nodeDefinitionId=_AA20EA8B-91E3-4171-8345-BFFBB103CD7F, nodeName=ScriptTask2, nodeType=ActionNode, nodeInstanceId=8ab42c0d-34e4-4cde-a640-2ac259b699ac, data={}],<br/>    ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:23 EST 2024, eventUser=null, eventType=2, processId=compensateAll, processVersion=1.0, processInstanceId=5e4847e5-1990-4d78-9368-c292d7f66c5f, connectionNodeInstanceId=_FF746A9A-D85F-4A83-918B-64808ACDE8E6, nodeDefinitionId=_8B1500BC-7458-4BFA-AB61-CC3AF6F375F9, nodeName=ScriptTask, nodeType=ActionNode, nodeInstanceId=efc7dcba-5aef-4907-ba32-3d423bc6817c, data={}],<br/>    ProcessInstanceNodeEventBody [eventDate=Mon Dec 02 03:14:23 EST 2024, eventUser=null, eventType=2, processId=compensateAll, processVersion=1.0, processInstanceId=5e4847e5-1990-4d78-9368-c292d7f66c5f, connectionNodeInstanceId=_66C7CEB0-2A49-4328-9E9D-0AD75C898486, nodeDefinitionId=_8AD8A66C-6327-4F26-BB2E-314283BED823, nodeName=Start, nodeType=StartNode, nodeInstanceId=18a58637-7546-409c-a917-0505e34af55b, data={}]]
</details>

pibizza commented on jbpm/jbpm-flow/src/test/java/org/jbpm/process/core/timer/CalendarBeanTest.java:
```diff
@@ -0,0 +1,295 @@
+        assertThat(errors).isNotEmpty();
+        String[] retrievedErrors = errors.toString().split("\n");
+        assertThat(retrievedErrors).hasSize(errorMessages.size());
+        assertThat(retrievedErrors).contains(retrievedErrors);
```
This is either wrong or meaningless.

gitgabrio commented:
> @Abhitocode hi, as I reviewed #3795 and this PR is related, I share feedback also on this PR.
> ### 01
> 
> As new potential contributor to runtimes and calendars, reading the code:
> 
> ```
> businessCal.rollCalendarAfterHolidays(calendar, false);
> ```
> 
> It is difficult to understand the `false`. I mean, without checking the actual method code, it is not clear what that false means. I understand that author of the code may not have this problem, but new contributors may be in similar situation as me.
> 
> What do you think about introducing two separate API methods:
> 
>     * `businessCal.rollCalendarAfterHolidays(calendar); // businessCal.rollCalendarAfterHolidays(calendar, false);`
> 
>     * `businessCal.rollCalendarAfterHolidaysAndResetTime(calendar); //businessCal.rollCalendarAfterHolidays(calendar, true);`
> 
> 
> or maybe change the boolean to some enum?
> 
>     * `businessCal.rollCalendarAfterHolidays(calendar, RESET_TIMER.NO);`
> 
> 
> ### 02
> 
> ```
>     @Test
>     void rollCalendarToNextWorkingDay() {
>         List<Integer> workingDays = IntStream.range(Calendar.MONDAY, Calendar.SATURDAY).boxed().toList();
>         List<Integer> weekendDays = Arrays.asList(Calendar.SATURDAY, Calendar.SUNDAY);
>         Properties config = new Properties();
>         config.setProperty(START_HOUR, "9");
>         config.setProperty(END_HOUR, "17");
>         config.setProperty(WEEKEND_DAYS, weekendDays.stream().map(String::valueOf).collect(Collectors.joining(",")));
>         BusinessCalendarImpl businessCal = BusinessCalendarImpl.builder().withCalendarBean(new CalendarBean(config)).build();
> 
>         workingDays.forEach(workingDay -> {
>             Calendar calendar = getCalendarAtExpectedWeekDay(workingDay);
>             businessCal.rollCalendarToNextWorkingDay(calendar, false);
>             assertThat(calendar.get(Calendar.DAY_OF_WEEK)).isEqualTo(workingDay);
>         });
>         weekendDays.forEach(weekendDay -> {
>             Calendar calendar = getCalendarAtExpectedWeekDay(weekendDay);
>             businessCal.rollCalendarToNextWorkingDay(calendar, false);
>             assertThat(calendar.get(Calendar.DAY_OF_WEEK)).isEqualTo(Calendar.MONDAY);
>         });
>     }
> ```
> 
> from reading this test ^, is `rollCalendarToNextWorkingDay` actually rolling? It seems we roll `workingDay` and after rolling we compare for equality with the original value. I would expect if we roll, that new value is `+1`. I mean I would expect sth like:
> 
> ```
> workingDays.forEach(workingDay -> {
>             Calendar calendar = getCalendarAtExpectedWeekDay(workingDay);
>             businessCal.rollCalendarToNextWorkingDay(calendar, false);
>             assertThat(calendar.get(Calendar.DAY_OF_WEEK)).isEqualTo(workingDay + 1);
>         });
> ```
> 
> Sorry if my feedback is not useful, or it proves my lack of knowledge about the domain. Thank you for your work.

HI @jomarko 

The scope of that test is to verify the behavior in tow different situations - and I agree it could be split in two.
Anyway, in the case of "workingDays", we have to _check that the calendar does not roll_, and that's why the assertion is that the `Calendar.DAY_OF_WEEK` does not change

yesamer commented on jbpm/jbpm-flow/src/test/java/org/jbpm/process/core/timer/BusinessCalendarImplTest.java:
```diff
@@ -18,419 +18,325 @@
-        Date result = businessCal.calculateBusinessTimeAsDate("6d4h80m");
-
-        assertThat(formatDate("yyyy-MM-dd HH:mm", result)).isEqualTo(expectedDate);
+    void rollCalendarAfterHolidays() {
```
@Abhitocode @gitgabrio It seems to me this is unrelated to the provided link (ticket 1651). Is that a typo or this is another bug?

Abhitocode replied:
@yesamer, It actually has a bug, fails at certain time. Hence disabled for now.

gitgabrio replied:
🤔 
@Abhitocode 
The test is related to `BusinessCalendarImpl.rollCalendarAfterHolidays`: why it has been disabled, and bound to nightly hour issue ?
If it fails, it could be:
1. wrong test setup
2. wrong implementation (bad)

Either case, IMO it has to be fixed....

Abhitocode replied:
@gitgabrio Let me check again and fix it

gitgabrio replied:
@Abhitocode 
Thanks for the quick fix.
But now I'm confused: why you disabled it, and then re-enabled without changing anything ? (maybe I'm missing something here) 🤔

gitgabrio commented on jbpm/jbpm-flow/src/test/resources/logback-test.xml:
```diff
@@ -29,6 +29,7 @@
   </appender>
 
   <logger name="org.jbpm" level="info"/>
+  <logger name="org.jbpm.process.core.timer" level="debug"/>
```
Fine for me to set the level to info

gitgabrio commented:
The CI:: License headers failure should be fixed by https://github.com/apache/incubator-kie-kogito-runtimes/pull/3802

gitgabrio commented:
@Abhitocode 
Could you please merge your branch with main, to cleanup the CI License header failure ?
Then, could you verify why downstream projects are failing ?
Last, could you please explain what happened with that test that has been disabled and then re-enabled ? We should avoid to introduce another flacky test around

@yesamer @pefernan ^ ^

Abhitocode commented:
@gitgabrio, Apologies for my incapability and mistake, I think I have a mistaken it with other test which was resolved. I checked the test again at multiple time periods, It did not fail. Hence I re-enabled it.

 Failures
1. incubator-kie-kogito-apps/data-index/data-index-service/data-index-service-mongodb/src/test/java/org/kie/kogito/index/service/graphql/MongoWebSocketSubscriptionIT.java#L1] -  unrelated test failure

2.  [Junit Report: apache_incubator-kie-kogito-examples/kogito-quarkus-examples/process-business-calendar-example/src/test/java/org/kie/kogito/calendar/CreditCardProcessIT.java#L76]
1 expectation failed. Response body doesn't match expectation. Expected: "[]" Actual: [{"id":"4d978234-db80-449b-8e6d-3121af

- I have tried to look into the issue,

```
2024-12-02 22:09:50,212 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Triggered node 'Verify Payment' for process 'BusinessCalendarCreditBill' (88cac26b-d3d9-47cc-b794-621a67119b88)
2024-12-02 22:09:50,231 INFO  [org.kie.kog.ser.job.imp.SignalProcessInstanceOnExpiredTimer] (executor-thread-2) Skipping Job 0b85c8c4-fdc2-4f01-a9dc-dfcd06ce5a19. There is no process for pid 88cac26b-d3d9-47cc-b794-621a67119b88
2024-12-02 22:09:50,269 INFO  [org.kie.kog.app.UserTasks] (executor-thread-1) Registering user task _C10EDD5D-0C5F-4ADB-9B85-28D3B9C7D2A7 with task name null
2024-12-02 22:09:50,460 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Workflow 'BusinessCalendarCreditBill' (88cac26b-d3d9-47cc-b794-621a67119b88) was started, now 'Active'
```

Here timer is actually triggered after 1 sec but skips the job saying no process for pid, couldn't find the exact reason.

gitgabrio commented:
@Abhitocode 
About the failing job, I do not see any useful logs, while, in this case, it is useful to have lot of them.
With those in place, we could better understand what "strange" may happen around 22:09, since it could be some flacky/time related situation

gitgabrio commented on jbpm/jbpm-flow/src/main/java/org/jbpm/process/core/timer/BusinessCalendarImpl.java:
```diff
@@ -218,102 +164,187 @@ public Date calculateBusinessTimeAsDate(String timeExpression) {
+        if (startHour < endHour) {
+            rollCalendarToDailyWorkingHour(toRoll, startHour, endHour);
+        } else {
+            rollCalendarToNightlyWorkingHour(toRoll, startHour, endHour);
```
@Abhitocode 
Until https://github.com/apache/incubator-kie-issues/issues/1651 is not done, it is needed to throw an exception here, or anyway to manage it, but that branch should not be executed

Abhitocode replied:
Okay, will remove it

gitgabrio replied:
@Abhitocode  - you should not remove it - you have to manage it!
As the code currently is, it is possible to get here with startHour > endHour, and so this use case has to properly be dealt - and the way it is to throw and log exception.

kie-ci3 commented:
**PR job** `#19` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-runtimes -u https://github.com/apache/incubator-kie-kogito-runtimes/pull/3788 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3788/19/display/redirect

**Test results:**
- PASSED: 3512
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3788/19/testReport/org.jbpm.process.core.timer/BusinessCalendarImplTest/rollCalendarAfterHolidays/">org.jbpm.process.core.timer.BusinessCalendarImplTest.rollCalendarAfterHolidays</a></summary>
<br/>expected: 343<br/> but was: 342
</details>

Abhitocode commented:
@gitgabrio, For the one failing Business Calendar example test, I tried to debug and found few possible reason,

1. With old Business Calendar implementation

Abhitocode commented:
@gitgabrio, After debugging the issue with the failing Business Calendar example test `CreditCardProcessIT`, I have identified the following possible reasons for the failure:

Logs for old business calendar implementation
```
2024-12-04 23:33:08,418 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Triggered node 'Start' for process 'BusinessCalendarCreditBill' (d037c870-78e7-430c-862a-cbfce2cddc2b)
2024-12-04 23:33:08,423 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Triggered node 'Process Credit Bill' for process 'BusinessCalendarCreditBill' (d037c870-78e7-430c-862a-cbfce2cddc2b)
2024-12-04 23:33:08,458 INFO  [org.kie.kog.cal.bil.CreditCardService] (executor-thread-1) Paying credit card
2024-12-04 23:33:08,496 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Variable 'creditCardDetails' changed value from: 'null', to: 'org.kie.kogito.calendar.bill.CreditCardDetails@4ee86809'
2024-12-04 23:33:08,517 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Triggered node 'Verify Payment' for process 'BusinessCalendarCreditBill' (d037c870-78e7-430c-862a-cbfce2cddc2b)
2024-12-04 23:33:08,642 INFO  [org.kie.kog.app.UserTasks] (executor-thread-1) Registering user task _C10EDD5D-0C5F-4ADB-9B85-28D3B9C7D2A7 with task name null
2024-12-04 23:33:08,868 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Workflow 'BusinessCalendarCreditBill' (d037c870-78e7-430c-862a-cbfce2cddc2b) was started, now 'Active'
2024-12-04 23:33:37,037 INFO  [org.kie.kog.ser.job.imp.SignalProcessInstanceOnExpiredTimer] (executor-thread-1) Job 3a4169ba-4aff-4981-8a8c-751c1c7fc061 started
2024-12-04 23:33:37,055 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Triggered node 'Cancel Payment' for process 'BusinessCalendarCreditBill' (d037c870-78e7-430c-862a-cbfce2cddc2b)
2024-12-04 23:33:37,057 INFO  [org.kie.kog.cal.bil.CreditCardService] (executor-thread-1) cancelling bill
2024-12-04 23:33:37,059 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Variable 'creditCardDetails' changed value from: 'org.kie.kogito.calendar.bill.CreditCardDetails@4ee86809', to: 'org.kie.kogito.calendar.bill.CreditCardDetails@4ee86809'
2024-12-04 23:33:37,061 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Triggered node 'End' for process 'BusinessCalendarCreditBill' (d037c870-78e7-430c-862a-cbfce2cddc2b)
2024-12-04 23:33:37,194 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Workflow 'BusinessCalendarCreditBill' (d037c870-78e7-430c-862a-cbfce2cddc2b) completed

```

Logs for new Business calendar implementation
```
2024-12-04 23:43:19,392 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Triggered node 'Start' for process 'BusinessCalendarCreditBill' (cd7cf39a-c2d3-4840-a69b-320667aa4313)
2024-12-04 23:43:19,394 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Triggered node 'Process Credit Bill' for process 'BusinessCalendarCreditBill' (cd7cf39a-c2d3-4840-a69b-320667aa4313)
2024-12-04 23:43:19,400 INFO  [org.kie.kog.cal.bil.CreditCardService] (executor-thread-1) Paying credit card
2024-12-04 23:43:19,408 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Variable 'creditCardDetails' changed value from: 'null', to: 'org.kie.kogito.calendar.bill.CreditCardDetails@406f70e1'
2024-12-04 23:43:19,409 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Triggered node 'Verify Payment' for process 'BusinessCalendarCreditBill' (cd7cf39a-c2d3-4840-a69b-320667aa4313)
2024-12-04 23:43:19,418 INFO  [org.kie.kog.ser.job.imp.SignalProcessInstanceOnExpiredTimer] (executor-thread-2) Skipping Job 4d1501de-fe10-4635-907c-cc9491f557be. There is no process for pid cd7cf39a-c2d3-4840-a69b-320667aa4313
2024-12-04 23:43:19,427 INFO  [org.kie.kog.app.UserTasks] (executor-thread-1) Registering user task _C10EDD5D-0C5F-4ADB-9B85-28D3B9C7D2A7 with task name null
2024-12-04 23:43:19,554 INFO  [org.kie.kog.qua.pro.dev.DevModeWorkflowLogger] (executor-thread-1) Workflow 'BusinessCalendarCreditBill' (cd7cf39a-c2d3-4840-a69b-320667aa4313) was started, now 'Active'

```

1. Difference in Thread Execution for Job Processing

Comparing the logs from the old and new Business Calendar implementation there is difference in thread execution:

- Old Implementation:

The **SignalProcessInstanceOnExpiredTimer** is executed on `executor-thread-1` after registering the user task. Further Cancel Payment node is triggered and the test passes.

- New Implementation:

The **SignalProcessInstanceOnExpiredTimer** is executed on `executor-thread-2` before registering the user task. This results in the job being skipped with the log:
“Skipping Job, There is no process for pid”. Consequently, the Cancel Payment node is not triggered, and the test fails.

This change in thread execution order may be causing the process to miss key events. However, I'm not sure why this change in thread behavior is occurring.

2. Another difference is evaluating expression for calculateBusinessTimeAsDate("1s").getTime() - getCurrentTime()

- Old Implementation:
calculateBusinessTimeAsDate("1s").getTime() - getCurrentTime() evaluates to a positive value (e.g., 997ms), timer functions as expected and test passes

- New Implementation:

The same expression evaluates to a negative value (e.g., -3504009ms), which may becausing the timer to behave incorrectly and skip the job execution.

These could be the reasons for test failure, however I'm not exactly sure and couldn't find exact solution. I would need help solving this.

@martinweiler @pefernan, couldn't find anything beyond the above reasons, any suggestions/solutions would be helpful.

kie-ci3 commented:
**PR job** `#21` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>



build-chain build full_downstream  -f 'https://raw.githubusercontent.com/${AUTHOR:apache}/incubator-kie-kogito-pipelines/${BRANCH:main}/.ci/buildchain-config-pr-cdb.yaml' -o 'bc' -p apache/incubator-kie-kogito-runtimes -u https://github.com/apache/incubator-kie-kogito-runtimes/pull/3788 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3788/21/display/redirect

**Test results:**
- PASSED: 3513
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://ci-builds.apache.org/job/KIE/job/kogito/job/main/job/pullrequest_jobs/job/kogito-runtimes-pr/job/PR-3788/21/testReport/org.kie.kogito.codegen.tests/PublishEventIT/testServiceTaskProcessWithError/">org.kie.kogito.codegen.tests.PublishEventIT.testServiceTaskProcessWithError</a></summary>
<br/>Expected size: 1 but was: 2 in:<br/>[ProcessInstanceStateDataEvent {specVersion=1.0, id='528dbbd9-0cbe-4d19-bfa5-0643e1315c94', source=http://myhost/TimerProcess, type='ProcessInstanceStateDataEvent', time=2024-12-05T09:35:30.403030430-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=ProcessInstanceStateEventBody [eventDate=Thu Dec 05 09:35:30 EST 2024, eventUser=null, eventType=1, processId=defaultPackage.TimerProcess, processVersion=1.0, processType=BPMN, processInstanceId=47b98999-f3c7-4f8d-b5e9-224df514df8d, businessKey=null, processName=TimerProcess, parentInstanceId=null, rootProcessId=null, rootProcessInstanceId=null, state=1, roles=null], kogitoProcessInstanceId='47b98999-f3c7-4f8d-b5e9-224df514df8d', kogitoRootProcessInstanceId='null', kogitoProcessId='defaultPackage.TimerProcess', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitoproctype=BPMN, kogitoprocinstanceid=47b98999-f3c7-4f8d-b5e9-224df514df8d, kogitoprocist=1, kogitoprocversion=1.0, kogitoprocid=defaultPackage.TimerProcess, kogitoaddons=test}},<br/>    ProcessInstanceStateDataEvent {specVersion=1.0, id='c5b5c68f-9af4-4ad6-b964-0a16c35bb199', source=http://myhost/ServiceProcessDifferentOperations, type='ProcessInstanceStateDataEvent', time=2024-12-05T09:35:30.406547943-05:00, subject='null', dataContentType='application/json', dataSchema=null, data=ProcessInstanceStateEventBody [eventDate=Thu Dec 05 09:35:30 EST 2024, eventUser=null, eventType=2, processId=ServiceProcessDifferentOperations, processVersion=1.0, processType=BPMN, processInstanceId=45f00adb-a2ee-4e7f-ba32-b2680fc9835b, businessKey=null, processName=Service Process, parentInstanceId=null, rootProcessId=null, rootProcessInstanceId=null, state=2, roles=null], kogitoProcessInstanceId='45f00adb-a2ee-4e7f-ba32-b2680fc9835b', kogitoRootProcessInstanceId='null', kogitoProcessId='ServiceProcessDifferentOperations', kogitoRootProcessId='null', kogitoAddons='test', kogitoIdentity='null', extensionAttributes={kogitoproctype=BPMN, kogitoprocinstanceid=45f00adb-a2ee-4e7f-ba32-b2680fc9835b, kogitoprocist=2, kogitoprocversion=1.0, kogitoprocid=ServiceProcessDifferentOperations, kogitoaddons=test}}]
</details>

gitgabrio commented:
@yesamer @pefernan @pibizza @martinweiler @jomarko 
Could you please review code again ? IINW all comments have been addressed.

jomarko commented:
sure, will do re-review soon

martinweiler reviewed (changes requested):
@Abhitocode this is generating too much logging output at DEBUG level. While I understand that some of the logging could be helpful to troubleshoot any miscalculations, I think we should move these to TRACE. DEBUG level logging is widely used to troubleshoot any workflow execution related issue, and I don't think we need this level of detail for all these cases.

Please check out https://gist.github.com/martinweiler/ae6a24eef68968794de86d8bc77d0b67 for a suggestion on how to change the logging. Thanks!

martinweiler reviewed (approved):
LGTM - thank you @Abhitocode and @gitgabrio and everyone involved

Abhitocode commented on jbpm/jbpm-flow/src/main/java/org/jbpm/process/core/timer/BusinessCalendarImpl.java:
```diff
@@ -218,244 +164,322 @@ public Date calculateBusinessTimeAsDate(String timeExpression) {
 
-                    Calendar tmp = new GregorianCalendar();
-                    tmp.setTime(holiday.getTo());
+                    Calendar firstWorkingHourTmp = new GregorianCalendar();
```
If the time at holiday ends is also the same time working day starts so I thought it would be relevant to name first working day hour. Your modifications makes more sense. Thank you.

pefernan reviewed (approved):
Great work guys! Thank you very much for this refactor 🚀

jomarko commented on jbpm/jbpm-flow/src/test/java/org/jbpm/process/core/timer/BusinessCalendarImplTest.java:
```diff
@@ -18,491 +18,367 @@
 
     @Test
-    public void testCalculateTimeDaysHoursMinutesHolidays() {
+    void calculateBusinessTimeAsDateBeforeWorkingHourWithDelayFineGrained() {
```
I like the structure of this test `calculateBusinessTimeAsDateBeforeWorkingHourWithDelayFineGrained`. I can understand its goal even without deep knowwledge of `BusinessCalendar`ish code. 

Tests, that reduce the code repetition using `commonCalculateBusinessTimeAsDateAssertBetweenHours` are not easy to read and understand for me. My comment is, that more natural language === more smaller methods would help to understand the tests. However I am not going to block the PR as I am not sure if we have resources to address my point.

gitgabrio replied:
@jomarko 
I see your point, and we discussed the same with @pibizza some days ago.
We have created a [ticket](https://github.com/apache/incubator-kie-issues/issues/1661) to make the "common*" tests parametrized.
The only caveat is to to also provide a test "name" (for each entry set) to also identify the given test (beside the tested values).

jomarko commented on jbpm/jbpm-flow/src/test/java/org/jbpm/process/core/timer/CalendarBeanFactoryTest.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Test
+    void testCreateCalendarBean() {
+        // This test relies on src/test/resources/calendar.properties:
```
In the `CalendarBean` I think we have code, that handles also parsing of values maybe also keys in the properties file, maybe we should add also a test that would load `calendar_broken.prperties` what would contain some unexpected content (eg. business.end.hour=XXX), What do you think?

jomarko replied:
Ok, it is probably in CalendarBeanTest

yesamer commented:
@gitgabrio @Abhitocode I checked the PR (sorry for the delay) and it looks good to me, thank you for your effort.
I guess the related example in `kogito-example` should be updated according to those changes, right?

Abhitocode commented:
@Yeser, right, example should be updated. Will update it once done with enablement.
